src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,703 @@
+
+  // initialize loop_entry history vars;
+  {
+    // Process "loop_entry" history variables.
+    // We find and replace all "__CPROVER_loop_entry" in invariant.
+    replace_history_parameter(
+      symbol_table,
+      invariant,
+      history_var_map,
+      loop_head_location,
+      language_mode,
+      pre_loop_head_instrs,
+      ID_loop_entry);
+  }
+
+  // entered_loop = false
+  {
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(entered_loop, loop_head_location));
+    pre_loop_head_instrs.add(
+      goto_programt::make_assignment(entered_loop, false_exprt{}));
+  }
+
+  // initial_invariant = <loop_invariant>;
+  {
+    goto_convertt converter(symbol_table, log.get_message_handler());
+    // Create a snapshot of the invariant so that we can check the base case,
+    // if the loop is not vacuous and must be abstracted with contracts.
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(initial_invariant, loop_head_location));
+
+    // Although the invariant at this point will not have side effects,
+    // it is still a C expression, and needs to be "goto_convert"ed.
+    // Note that this conversion may emit many GOTO instructions.
+    code_assignt initial_invariant_assignment{initial_invariant, invariant};
+    initial_invariant_assignment.add_source_location() = loop_head_location;
+    converter.goto_convert(
+      initial_invariant_assignment, pre_loop_head_instrs, language_mode);
+  }
+
+  {
+    // Create a temporary variable to track base case vs inductive case
+    // instrumentation of the loop.
+    // in_base_case = true;
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(in_base_case, loop_head_location));
+    pre_loop_head_instrs.add(
+      goto_programt::make_assignment(in_base_case, true_exprt{}));
+  }
+
+  {
+    // Create and populate the write set.
+    // DECL loop_write_set
+    // DECL addr_of_loop_write_set
+    // ASSIGN write_set_ptr := address_of(write_set)
+    // CALL __CPROVER_contracts_write_set_create(write_set_ptr,
+    //        contracts_assigns_size, contracts_assigns_frees_size,
+    //        assume_require_ctx, assert_require_ctx, assume_ensures_ctx,
+    //        assert_ensures_ctx, allow_allocate, allow_deallocate);
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(loop_write_set, loop_head_location));
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(addr_of_loop_write_set, loop_head_location));
+    pre_loop_head_instrs.add(goto_programt::make_assignment(
+      addr_of_loop_write_set,
+      address_of_exprt(loop_write_set),
+      loop_head_location));
+
+    code_function_callt call = library.write_set_create_call(
+      addr_of_loop_write_set,
+      from_integer(assigns.size(), size_type()),
+      from_integer(0, size_type()),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      false_exprt(),
+      loop_head_location);
```

Comment:
That's a lot of `false_exprt` and also that `from_integer(0, ...`: should there perhaps be a variant of `write_set_create_call` that fills all those in?